File: python/features/steps/users_steps.py
"""Step definitions for current user resolution."""

from __future__ import annotations

import os

from behave import given, then, when

from kanbus.users import get_current_user


@given('KANBUS_USER is set to "{value}"')
def given_kanbus_user_set(context: object, value: str) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides["KANBUS_USER"] = value
    context.environment_overrides = overrides


@given("KANBUS_USER is unset")
def given_kanbus_user_unset(context: object) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides.pop("KANBUS_USER", None)
    context.environment_overrides = overrides


@given('USER is set to "{value}"')
def given_user_set(context: object, value: str) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides["USER"] = value
    context.environment_overrides = overrides


@given("USER is unset")
def given_user_unset(context: object) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides.pop("USER", None)
    context.environment_overrides = overrides


@when("I resolve the current user")
def when_resolve_current_user(context: object) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    original_kanbus_user = os.environ.get("KANBUS_USER")
    original_user = os.environ.get("USER")
    try:
        if "KANBUS_USER" in overrides:
            os.environ["KANBUS_USER"] = overrides["KANBUS_USER"]
        else:
            os.environ.pop("KANBUS_USER", None)
        if "USER" in overrides:
            os.environ["USER"] = overrides["USER"]
        else:
            os.environ.pop("USER", None)
        context.current_user = get_current_user()
    finally:
        if original_kanbus_user is None:
            os.environ.pop("KANBUS_USER", None)
        else:
            os.environ["KANBUS_USER"] = original_kanbus_user
        if original_user is None:
            os.environ.pop("USER", None)
        else:
            os.environ["USER"] = original_user


@then('the current user should be "{value}"')
def then_current_user_is(context: object, value: str) -> None:
    current_user = getattr(context, "current_user", None)
    assert current_user == value

File: rust/features/steps/users_steps.rs
use cucumber::{given, then, when};

use kanbus::users::get_current_user;

use crate::step_definitions::initialization_steps::KanbusWorld;

fn capture_original_env(world: &mut KanbusWorld) {
    if world.original_kanbus_user.is_none() {
        world.original_kanbus_user = Some(std::env::var("KANBUS_USER").ok());
    }
    if world.original_user_env.is_none() {
        world.original_user_env = Some(std::env::var("USER").ok());
    }
}

#[given(expr = "KANBUS_USER is set to {string}")]
fn given_kanbus_user_set(world: &mut KanbusWorld, value: String) {
    capture_original_env(world);
    std::env::set_var("KANBUS_USER", value);
}

#[given("KANBUS_USER is unset")]
fn given_kanbus_user_unset(world: &mut KanbusWorld) {
    capture_original_env(world);
    std::env::remove_var("KANBUS_USER");
}

#[given(expr = "USER is set to {string}")]
fn given_user_set(world: &mut KanbusWorld, value: String) {
    capture_original_env(world);
    std::env::set_var("USER", value);
}

#[given("USER is unset")]
fn given_user_unset(world: &mut KanbusWorld) {
    capture_original_env(world);
    std::env::remove_var("USER");
}

#[when("I resolve the current user")]
fn when_resolve_current_user(world: &mut KanbusWorld) {
    world.current_user = Some(get_current_user());
}

#[then(expr = "the current user should be {string}")]
fn then_current_user_is(world: &mut KanbusWorld, value: String) {
    assert_eq!(world.current_user.as_deref(), Some(value.as_str()));
}
